src/Connection.php

```diff
@@ -1318,15 +1317,6 @@ private function bindParameters(DriverStatement $stmt, array $params, array $typ
                     $type                  = $types[$key];
                     [$value, $bindingType] = $this->getBindingInfo($value, $type);
                 } else {
-                    if (array_key_exists($key, $types)) {
-                        Deprecation::trigger(
-                            'doctrine/dbal',
-                            'https://github.com/doctrine/dbal/pull/5550',
-                            'Using NULL as prepared statement parameter type is deprecated.'
-                                . 'Omit or use Parameter::STRING instead'
-                        );
-                    }
```

Comment:
This basically means that the deprecation had no effect: It is still possible to pass `null` as type.

Reply:
This is not enforced at runtime because connection methods accept types as an array. If I misunderstand your point, could you illustrate it with a code snippet?

Reply:
> This is not enforced at runtime because connection methods accept types as an array.

This is okay, but then it did not make sense that we triggered that deprecation. The caller is warned about a call that continues to work.

> If I misunderstand your point, could you illustrate it with a code snippet?

If I understood the deprecation correctly, it is triggered on calls like this:

```php
$connection->executeQuery(
    sql: $sql,
    params: ['myParam' => 'foo'],
    types: ['myParam' => null]
);
```

The change you are suggesting is to remove the deprecation trigger without introducing a breaking change. The call above remains intact.

> Passing NULL to the wrapper will likely result in `getBindingInfo()` returning NULL for `$bindingType` which in turn will cause an exception in the driver because the driver-level `bind*()` methods don't accept NULL.

No, passing `null` has the same effect as omitting that particular parameter in the `$types` array: We fall back to string. This is literally what the line after this block does.

My suggesting would be to either:

* Revert the deprecation on 3.4.x or
* Throw if we encounter `null` in this array.

Reply:
I see. You're right. Should we just replace the`isset()` in the below condition with `array_key_exists()`? https://github.com/doctrine/dbal/blob/0cc7adc22ea3d81a425a48463044931e95693704/src/Connection.php#L1317-L1320

Reply:
We could do that. That would trigger a `TypeError` on the call to `getBindingInfo()` because you've removed `null` from the union there.